apps/opencs/view/render/terraintexturemode.cpp

```diff
@@ -0,0 +1,214 @@
+// To-do: Getting Texture Id and Texture Filenames to base class Terraintexturemode
+// To-do: Better data handling options for mBrushTexture
+// To-do: loading texture bitmaps from virtual file system (vfs) for texture overlay icon
+
+#include "terraintexturemode.hpp"
+#include "editmode.hpp"
+
+#include <string>
+
+#include <QWidget>
+#include <QIcon>
+#include <QSpinBox>
+#include <QGroupBox>
+#include <QSlider>
+#include <QEvent>
+#include <QDropEvent>
+#include <QButtonGroup>
+#include <QHBoxLayout>
+#include <QVBoxLayout>
+#include <QDragEnterEvent>
+#include <QDrag>
+
+#include "../widget/modebutton.hpp"
+#include "../widget/scenetoolbar.hpp"
+
+#include "../../model/world/universalid.hpp"
+#include "../../model/world/tablemimedata.hpp"
+
+#include "pagedworldspacewidget.hpp"
+
+
+CSVRender::BrushSizeControls::BrushSizeControls(const QString &title, QWidget *parent)
+    : QGroupBox(title, parent)
+{
+    brushSizeSlider = new QSlider(Qt::Horizontal);
+    brushSizeSlider->setTickPosition(QSlider::TicksBothSides);
+    brushSizeSlider->setTickInterval(10);
+    brushSizeSlider->setSingleStep(1);
+
+    brushSizeSpinBox = new QSpinBox;
+    brushSizeSpinBox->setRange(1, 100);
+    brushSizeSpinBox->setSingleStep(1);
+
+    layoutSliderSize = new QHBoxLayout;
+    layoutSliderSize->addWidget(brushSizeSlider);
+    layoutSliderSize->addWidget(brushSizeSpinBox);
+
+    connect(brushSizeSlider, SIGNAL(valueChanged(int)), brushSizeSpinBox, SLOT(setValue(int)));
+    connect(brushSizeSpinBox, SIGNAL(valueChanged(int)), brushSizeSlider, SLOT(setValue(int)));
+
+    setLayout(layoutSliderSize);
+}
+
+CSVRender::TextureBrushButton::TextureBrushButton (const QIcon & icon, const QString & text, QWidget * parent)
+    : QPushButton(icon, text, parent)
+{
+}
+
+CSVRender::TextureBrushWindow::TextureBrushWindow(WorldspaceWidget *worldspaceWidget, QWidget *parent)
+    : QFrame(parent, Qt::Popup), mWorldspaceWidget (worldspaceWidget)
+{
+    mBrushTextureLabel = "Brush: " + mBrushTexture;
+    selectedBrush = new QLabel(QString::fromUtf8(mBrushTextureLabel.c_str()), this);
+
+    const std::string& iconPoint = ":scenetoolbar/brush-point";
+    const std::string& iconSquare = ":scenetoolbar/brush-square";
+    const std::string& iconCircle = ":scenetoolbar/brush-circle";
+    const std::string& iconCustom = ":scenetoolbar/brush-custom";
+
+    TextureBrushButton *buttonPoint = new TextureBrushButton(QIcon (QPixmap (iconPoint.c_str())), "", this);
```

Comment:
That is not how you convert between std::string and QString (at least not in Qt4). Messes up the encoding. There is an fromUtf8 function.

Reply:
Hm. Okay. I changed it to QString::fromStdString as QString::fromUtf8 takes only const char *str.